Mathlib/CategoryTheory/Sites/EffectiveEpimorphic.lean

```diff
@@ -0,0 +1,433 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+
+import Mathlib.CategoryTheory.Sites.Sieves
+import Mathlib.CategoryTheory.Limits.Shapes.KernelPair
+
+/-!
+
+# Effective epimorphisms
+
+We define the notion of effective epimorphic (pre)sieves, morphisms and family of morphisms
+and provide some API for relating the three notions.
+
+More precisely, if `f` is a morphism, then `f` is an effective epi if and only if the sieve
+it generates is effective epimorphic; see `CategoryTheory.Sieve.effectiveEpimorphic_singleton`.
+The analogous statement for a family of morphisms is in the theorem
+`CategoryTheory.Sieve.effectiveEpimorphic_family`.
+
+We have defined the notion of effective epi for morphisms and families of morphisms in such a
+way that avoids requiring the existence of pullbacks. However, if the relevant pullbacks exist
+then these definitions should be equivalent (project: formalize this!).
+See [nlab: *Effective Epimorphism*](https://ncatlab.org/nlab/show/effective+epimorphism) and
+[Stacks 00WP](https://stacks.math.columbia.edu/tag/00WP) for the standard definitions.
+
+## References
+- [Elephant]: *Sketches of an Elephant*, P. T. Johnstone: C2.1, Example 2.1.12.
+- [nlab: *Effective Epimorphism*](https://ncatlab.org/nlab/show/effective+epimorphism) and
+- [Stacks 00WP](https://stacks.math.columbia.edu/tag/00WP) for the standard definitions.
+
+-/
+
+namespace CategoryTheory
+
+open Limits
+
+variable {C : Type _} [Category C]
+
+/-- A sieve is effective epimorphic if the associated cocone is a colimit cocone. -/
+def Sieve.EffectiveEpimorphic {X : C} (S : Sieve X) : Prop :=
+  Nonempty (IsColimit (S : Presieve X).cocone)
+
+/-- A presieve is effective epimorphic if the cocone assocaited to the sieve it generates
+is a colimit cocone. -/
+abbrev Presieve.EffectiveEpimorphic {X : C} (S : Presieve X) : Prop :=
+  (Sieve.generate S).EffectiveEpimorphic
+
+/--
+The sieve of morphisms which factor through a given morphism `f`.
+This is equal to `Sieve.generate (Presieve.singleton f)`, but has
+more convenient definitional properties.
+-/
+def Sieve.generateSingleton {X Y : C} (f : Y ⟶ X) : Sieve X where
+  arrows Z := { g | ∃ (e : Z ⟶ Y), e ≫ f = g }
+  downward_closed := by
+    rintro W Z g ⟨e,rfl⟩ q
+    refine ⟨q ≫ e, by simp⟩
```

Comment:
Should this replace the other construction after the port? If so, please leave a porting note.

Reply:
I don't think so. `Sieve.generate` can take an atrbitrary presieve as input, and this just applies to presieves of the form `Presieve.singleton f`.

Reply:
Aah, of course. But it should maybe move at some point?